File: ios/Sources/RynBridgeTranslation/TranslationModule.swift
import Foundation
import RynBridge

public struct TranslationModule: BridgeModule, Sendable {
    public let name = "translation"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: TranslationProvider) {
        actions = [
            "translate": { payload in
                guard let text = payload["text"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: text")
                }
                guard let source = payload["source"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: source")
                }
                guard let target = payload["target"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: target")
                }
                let result = try await provider.translate(text: text, source: source, target: target)
                return result.toPayload()
            },
            "translateBatch": { payload in
                guard let textsArray = payload["texts"]?.arrayValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: texts")
                }
                let texts = textsArray.compactMap { $0.stringValue }
                guard let source = payload["source"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: source")
                }
                guard let target = payload["target"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: target")
                }
                let result = try await provider.translateBatch(texts: texts, source: source, target: target)
                return result.toPayload()
            },
            "detectLanguage": { payload in
                guard let text = payload["text"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: text")
                }
                let result = try await provider.detectLanguage(text: text)
                return result.toPayload()
            },
            "getSupportedLanguages": { _ in
                let result = try await provider.getSupportedLanguages()
                return result.toPayload()
            },
            "downloadModel": { payload in
                guard let language = payload["language"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: language")
                }
                let result = try await provider.downloadModel(language: language)
                return result.toPayload()
            },
            "deleteModel": { payload in
                guard let language = payload["language"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: language")
                }
                let result = try await provider.deleteModel(language: language)
                return result.toPayload()
            },
            "getDownloadedModels": { _ in
                let result = try await provider.getDownloadedModels()
                return result.toPayload()
            },
        ]
    }
}

File: android/translation/src/main/kotlin/com/devryner/rynbridge/translation/TranslationModule.kt
package com.devryner.rynbridge.translation

import com.devryner.rynbridge.core.*

class TranslationModule(provider: TranslationProvider) : BridgeModule {

    override val name = "translation"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "translate" to { payload ->
            val text = payload["text"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: text")
            val source = payload["source"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: source")
            val target = payload["target"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: target")
            val result = provider.translate(text, source, target)
            result.toPayload()
        },
        "translateBatch" to { payload ->
            val texts = payload["texts"]?.arrayValue
                ?.mapNotNull { it.stringValue }
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: texts")
            val source = payload["source"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: source")
            val target = payload["target"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: target")
            val result = provider.translateBatch(texts, source, target)
            result.toPayload()
        },
        "detectLanguage" to { payload ->
            val text = payload["text"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: text")
            val result = provider.detectLanguage(text)
            result.toPayload()
        },
        "getSupportedLanguages" to { _ ->
            val result = provider.getSupportedLanguages()
            result.toPayload()
        },
        "downloadModel" to { payload ->
            val language = payload["language"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: language")
            val result = provider.downloadModel(language)
            result.toPayload()
        },
        "deleteModel" to { payload ->
            val language = payload["language"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: language")
            val result = provider.deleteModel(language)
            result.toPayload()
        },
        "getDownloadedModels" to { _ ->
            val result = provider.getDownloadedModels()
            result.toPayload()
        }
    )
}
